doc/design/subscription.md

```diff
@@ -0,0 +1,175 @@
+| P1        | 11        | B      | | P2        | 22        | G      |
+| P1        | 21        | C      | | P2        | 24        | H      |
+| P1        | 31        | D      | | P2        | 40        | I      |
+| P1        | 41        | E      | | P2        | 42        | J      |
+
+Let's assume the subscriber subscribes to the service.
+
+#### Subscription with _HistorySize_ 0
+
+The subscriber will not be provided with samples from the history.
+
+#### Subscription with _HistorySize_ 1
+
+The subscriber cache contains the one sample ordered after _DeliveryTime_.
+
+| Publisher | Timepoint | Sample |
+|:---------:|:---------:|:------:|
+| P2        | 42        | J      |
+
+#### Subscription with _HistorySize_ 6
+
+The subscriber cache contains the six samples ordered after _DeliveryTime_.
+
+| Publisher | Timepoint | Sample |
+|:---------:|:---------:|:------:|
+| P2        | 42        | J      |
+| P1        | 41        | E      |
+| P2        | 40        | I      |
+| P1        | 31        | D      |
+| P2        | 24        | H      |
+| P2        | 22        | G      |
+
+#### Subscription with _HistorySize_ 20
+
+The subscriber cache contains the ten samples ordered after _DeliveryTime_.
+It contains only ten since there are no more samples available on publisher side.
+
+| Publisher | Timepoint | Sample |
+|:---------:|:---------:|:------:|
+| P2        | 42        | J      |
+| P1        | 41        | E      |
+| P2        | 40        | I      |
+| P1        | 31        | D      |
+| P2        | 24        | H      |
+| P2        | 22        | G      |
+| P1        | 21        | C      |
+| P2        | 20        | F      |
+| P1        | 11        | B      |
+| P1        | 1         | A      |
+
+### Subscribing with concurrent send
+
+Let's assume publisher P1 would like to deliver the sample K at timepoint
+52 while the subscriber requested subsription. Furthermore, the subscriber
+requests always a _HistorySize_ of 2.
+
+All the listed cases should behave identical if there is more then one
+concurrent publish in progress.
+
+#### Publisher before _DeliveryTime_
+
+The subscriber cache contains two samples.
+
+| Publisher | Timepoint | Sample |
+|:---------:|:---------:|:------:|
+| P2        | 42        | J      |
+| P1        | 41        | E      |
+
+The sample K will be delivered after successful subscription.
+
+#### Publisher after _DeliveryTime_
+
+The subscriber cache contains two samples.
+
+| Publisher | Timepoint | Sample |
+|:---------:|:---------:|:------:|
+| P1        | 51        | K      |
+| P2        | 42        | J      |
+
+The sample K will not be delivered by the publisher since it is already
+present in the cache.
+
+### Subscribing After Publisher Stops Offering
+
+In this case we assume that publisher P1 has stopped offering the service
+before the subscription was requested.
+
+#### Subscription with _HistorySize_ 4
+
+The subscriber cache contains the four samples ordered after _DeliveryTime_.
```

Comment:
Yes, the subscriber is not subscribing to a publisher it is subscribing to a service. And the user requested the last 4 samples send in the name of that service.

Let's say you request a history of size two and the last four samples would be A, B, C and D. And now the publisher stops offering (maybe even not willingly, maybe something unexpected happened), and now you receive a history with holes like A and D.
This is quite error prone and not robust and as a user I would really wish to get the last 4 samples from the service and not depending on what publisher did send what.

Also the name history size would then be misleading since it is not the actually history ... it is again the history which was written by the winners. Again! ... ;)

Reply:
Yes, it probably has to be done at the reader side or at service level. Not only because offering may be stopped, but also because we could have multiple publishers for the same topic. Which data shall we receive then? A cache at the reader side would handle this but require needless replication for all readers. This is also bad since in the order of elements will differ between readers due to the nature of concurrent publishing.

A centralized cache for the topic would be the solution, but then we would a multi-writer multi-reader cache (the MPMC FIFO is not sufficient as pop is destructive). For now I can only offer a relatively easy to implement single writer multi reader cache as in https://github.com/eclipse-iceoryx/iceoryx/issues/857 which would solve the problem if the cache exists per problem (as it currently is).